### ISSUE TITLE
Test: "create and dispose" in TcpRPC hangs indefinitely

### ISSUE THREAD

**Issue by lostinplace** [c80k/capnproto-dotnetcore#3]
the test seems to hang indefinitely

![image](https://user-images.githubusercontent.com/645413/59873947-1edb8f00-936b-11e9-844c-e418a87d16c7.png)

![image](https://user-images.githubusercontent.com/645413/59874118-8560ad00-936b-11e9-8f1c-fd969b33ba86.png)

![image](https://user-images.githubusercontent.com/645413/59874160-9c9f9a80-936b-11e9-9f94-fa39ce3dbb94.png)

![image](https://user-images.githubusercontent.com/645413/59874204-b80aa580-936b-11e9-96e0-221927111a17.png)

**Comment by c80k** [comment]
Thanks for the finding and sorry for the answering delay. I'm currently in holidays and have only limited options here (especially no machine for debugging this issue). If you like to investigate this in more detail: There are 3 threads involved whose call stacks might be interesting. They all have meaningful names such that you can identify them amongst the many existing threads:
- The "Test" is the test case main thread. If the TC hangs, this thread must hang, too. Where is it stuck?
- An "RPC Server" thread is created for each incoming connection (as opposed to the "TCP Acceptor" thread, which exists only once) and runs the server-side message loop. There should be exactly one. Is it there?
- An "RPC Client" thread is created for each outgoing connection, running the client-side message loop. Is it there?

**Comment by c80k** [comment]
Another suggestion: Maybe you are firewalled. Is the TCP port permitted on your system?

**Comment by lostinplace** [comment]
This ultimately came down to a macOS difference.  The serverclientpair setup fails because macOS and linux systems won't allow multiple listeners on the same port, unlike windows which doesn't care.  The tests should be rewritten to not rely all one port, but an alternative approach to the problem is to not run the tests on linux or macOS

**Comment by c80k** [comment]
The tests are expected to run sequentially. Once a TC is finished it should also release the TCP listener port. So it should never happen that we have two listeners on the same port. Does the TCP acceptor thread not shut down properly with server disposal? Or is there some different (completely unrelated) server which listens on the same port (by incidence)?

**Comment by lostinplace** [comment]
I didn't see any logic that would release access to the port after transmission was finished.  It's definitely not another process grabbing the port and the specific exception is triggered by the .NET framework on the access attempt.

**Comment by lostinplace** [comment]
Even if the port were getting released, the likelihood of a deadlock is pretty high.  Maybe it needs a mutex?

**Comment by lostinplace** [comment]
just to confirm, the tcp listener for the desired port is established outside of the listen thread (in the constructor), and so is the TcpClient for the client

**Comment by c80k** [comment]
Correct. Both TcpRpcClient and TcpRpcServer implement IDisposable. Calling Dispose shall release all resources and also join all threads. So my expectation was that outside of the „using“ blocks, no TCP ports are allocated.

**Comment by c80k** [comment]
Correct. Both TcpRpcClient and TcpRpcServer implement IDisposable. Calling Dispose shall release all resources and also join all threads. So my expectation was that outside of the „using“ blocks, no TCP ports are allocated.

**Comment by lostinplace** [comment]
The problem appears to be that both the client and the server are making requests of the port at the same time, the server is calling `AcceptTcpClient()` which in turn relies on `TcpClient(this._serverSocket.Accept());` while the client is calling `ConnectAsync(host, port)`

To be clear, the exception is being thrown at client disposal time, when evaluating the `WhenConnected` task (only to trigger the recognition that there was contention for the port allocation)

**Comment by c80k** [comment]
Observed some tests failing sporadically on Windows platform and found another issue: When the connection breaks down for whatever reason (applies to both client and server side) the RpcEngine does not set pending questions to a faulted state. This might cause a question to hang indefinitely. This is addressed with commit eaecfda3 in net-standard-runtime-mixed branch.

The TcpRpcClient exception is probably due to the TIME_WAIT state (found this article: http://www.serverframework.com/asynchronousevents/2011/01/time-wait-and-its-design-implications-for-protocols-and-scalable-servers.html). Some TCs establish and close lots of TCP connections in a short amount of time. It seems that the OS runs out of ephemeral ports. This would be an explanation why setting ExclusiveAddressUse=false works around the problem.

**Comment by c80k** [comment]
Should be solved.
